File: src/Domain/Teams/ValueObjects/State.cs
using System.Collections.Generic;
using System.Collections.Immutable;
using Snaelro.Domain.Teams.Events;

namespace Snaelro.Domain.Teams.ValueObjects
{
    public class State
    {
        public bool Created { get; private set; }

        public string Name { get; private set; }

        public IImmutableList<string> Players { get; private set; }

        public State()
        {
            Players = new List<string>().ToImmutableList();
        }

        public void Apply(TeamCreated @event)
        {
            Created = true;
            Name = @event.Name;
        }

        public void Apply(PlayerAdded @event)
        {
            Players = Players.Add(@event.Name);
        }
    }
}

File: src/DomainF/Teams/State.fs
namespace Orleans.Tournament.Domain.Teams
open Orleans.Tournament.Domain.Teams.Events
open System

type TeamState =
    val mutable Id:Guid
    val mutable Created:bool
    val mutable Name:string
    val mutable Players:List<string>
    val mutable Tournaments:List<Guid>
    new() = {
        Id = Guid.Empty;
        Created = false;
        Name = String.Empty;
        Players = []
        Tournaments = []
    }
    member x.Apply (teamCreated : TeamCreated) =
        x.Created <- true;
        x.Id <- teamCreated.TeamId;
        x.Name <- teamCreated.Name;
    member x.Apply (playerAdded : PlayerAdded) =
        x.Players = playerAdded.Name :: x.Players
    member x.Apply (tournamentJoined : TournamentJoined) =
        x.Tournaments = tournamentJoined.TournamentId :: x.Tournaments;
